synapse/lib/cache.py

```diff
@@ -197,7 +197,7 @@ def __init__(self, maxsize=10000, onmiss=None):
         self.cache = {}
         self.onmiss = onmiss
         self.maxsize = maxsize
-        self.cachelock = threading.Lock()
+        self.cachelock = threading.RLock()
```

Comment:
( ie, when tag foo.bar.baz is newly created and the parents dont exist either )  I timed the perf difference, which goes from roughly 0.29 usec to 0.30 usec ( per lock request ) which is negligible for the simplicity benefit ;)